janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/service/GroupingAsyncQueryExecutionService.java

```diff
@@ -0,0 +1,484 @@
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.COLUMN_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.COLUMN_COLUMN_NAME;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.KEY_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.KEY_COLUMN_NAME;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.LIMIT_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.SLICE_END_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.SLICE_START_BINDING;
+import static org.janusgraph.diskstorage.cql.CQLKeyColumnValueStore.VALUE_COLUMN_NAME;
+
+public class GroupingAsyncQueryExecutionService implements AsyncQueryExecutionService {
+
+    private final ResultFiller<Map<StaticBuffer, CompletableFuture<EntryList>>, SliceQuery, TokenRangeKeyGroup> SINGLE_QUERY_WITH_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<StaticBuffer, CompletableFuture<EntryList>>, SliceQuery, List<StaticBuffer>> SINGLE_QUERY_WITHOUT_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<SliceQuery, Map<StaticBuffer, CompletableFuture<EntryList>>>, QueryGroups, TokenRangeKeyGroup> MULTI_QUERY_WITH_KEYS_GROUPING_FILLER;
+    private final ResultFiller<Map<SliceQuery, Map<StaticBuffer, CompletableFuture<EntryList>>>, QueryGroups, List<StaticBuffer>> MULTI_QUERY_WITHOUT_KEYS_GROUPING_FILLER;
+
+    private final CqlSession session;
+    private final AsyncCQLSingleKeySliceFunction cqlSingleKeySliceFunction;
+    private final AsyncCQLSingleKeyMultiColumnFunction cqlSingleKeyMultiColumnFunction;
+    private final AsyncCQLMultiKeySliceFunction cqlMultiKeySliceFunction;
+    private final AsyncCQLMultiKeyMultiColumnFunction cqlMultiKeyMultiColumnFunction;
+    private final boolean sliceGroupingAllowed;
+    private final int sliceGroupingLimit;
+    private final boolean keysGroupingAllowed;
+    private final int keysGroupingLimit;
+    private final int keysGroupingMin;
+
+    public GroupingAsyncQueryExecutionService(Configuration configuration,
+                                              CqlSession session,
+                                              String keyspaceName,
+                                              String edgeStoreTableName,
+                                              Function<Select, Select> addTTLFunction,
+                                              Function<Select, Select> addTimestampFunction,
+                                              CQLColValGetter singleKeyGetter,
+                                              CQLColValGetter multiKeysGetter,
+                                              ExecutorService executorService,
+                                              QueryBackPressure queryBackPressure) {
+        this.session = session;
+        sliceGroupingLimit = configuration.get(SLICE_GROUPING_LIMIT) < 1 ? 1 : configuration.get(SLICE_GROUPING_LIMIT);
```

Comment:
```suggestion
        sliceGroupingLimit = Math.min(configuration.get(SLICE_GROUPING_LIMIT), 1);
```

Reply:
I think it should be ‘max’ instead of ‘min’ here, but I agree that it looks better like you suggested. I will change it.

Reply:
Oh yeah it should be max

Reply:
Applied `Math.max`